File: frontend/src/api/backup.ts
/** API hooks for backup/restore (P4-21c). */

import { useQuery, useMutation } from '@tanstack/react-query'

// ── Types ────────────────────────────────────────────────────────────

export interface BackupEstimate {
  sample_bytes: number
  config_bytes: number
  reference_bytes: number
  total_without_ref_bytes: number
  total_with_ref_bytes: number
  total_without_ref_mb: number
  total_with_ref_mb: number
  sample_count: number
  reference_db_count: number
}

export interface BackupExportResult {
  job_id: string
  message: string
}

export interface BackupStatus {
  job_id: string
  status: 'pending' | 'running' | 'complete' | 'failed'
  progress_pct: number
  message: string
  error: string | null
  download_filename: string | null
}

// ── Query keys ───────────────────────────────────────────────────────

export const BACKUP_ESTIMATE_KEY = ['backup', 'estimate'] as const

// ── Fetch functions ─────────────────────────────────────────────────

async function fetchBackupEstimate(): Promise<BackupEstimate> {
  const res = await fetch('/api/backup/estimate')
  if (!res.ok) throw new Error(`Failed to fetch backup estimate: ${res.status}`)
  return res.json()
}

async function startBackupExport(includeReferenceDbs: boolean): Promise<BackupExportResult> {
  const res = await fetch('/api/backup/export', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ include_reference_dbs: includeReferenceDbs }),
  })
  if (!res.ok) throw new Error(`Failed to start backup: ${res.status}`)
  return res.json()
}

export async function fetchBackupStatus(jobId: string): Promise<BackupStatus> {
  const res = await fetch(`/api/backup/status/${jobId}`)
  if (!res.ok) throw new Error(`Failed to fetch backup status: ${res.status}`)
  return res.json()
}

// ── Hooks ───────────────────────────────────────────────────────────

export function useBackupEstimate() {
  return useQuery({
    queryKey: BACKUP_ESTIMATE_KEY,
    queryFn: fetchBackupEstimate,
    staleTime: 30 * 1000,
  })
}

export function useStartBackupExport() {
  return useMutation({
    mutationFn: (includeReferenceDbs: boolean) => startBackupExport(includeReferenceDbs),
  })
}

export function useBackupStatus(jobId: string | null) {
  return useQuery({
    queryKey: ['backup', 'status', jobId] as const,
    queryFn: () => fetchBackupStatus(jobId!),
    enabled: !!jobId,
    refetchInterval: (query) => {
      const status = query.state.data?.status
      if (status === 'complete' || status === 'failed') return false
      return 1000
    },
  })
}

File: backend/api/routes/backup.py
"""Backup export API routes (P4-21c).

Endpoints:
    GET  /api/backup/estimate            — Estimate archive size before export
    POST /api/backup/export              — Start background export as Huey task
    GET  /api/backup/status/{job_id}     — Poll export job status
    GET  /api/backup/download/{filename} — Download completed archive
"""

from __future__ import annotations

from pathlib import Path

import structlog
from fastapi import APIRouter, HTTPException
from fastapi.responses import FileResponse
from pydantic import BaseModel

from backend.config import get_settings

logger = structlog.get_logger(__name__)

router = APIRouter(prefix="/backup", tags=["backup"])

# Reference DB filenames that can optionally be included in the archive.
# Shared with the Huey task via import.
REFERENCE_DB_FILES = (
    "clinvar.db",
    "vep_bundle.db",
    "gnomad_af.db",
    "dbnsfp.db",
    "encode_ccres.db",
    "reference.db",
)


# ── Response models ──────────────────────────────────────────────────


class BackupEstimateResponse(BaseModel):
    """Estimated archive size before export."""

    sample_bytes: int
    config_bytes: int
    reference_bytes: int
    total_without_ref_bytes: int
    total_with_ref_bytes: int
    total_without_ref_mb: float
    total_with_ref_mb: float
    sample_count: int
    reference_db_count: int


class BackupExportRequest(BaseModel):
    """Request to start a backup export."""

    include_reference_dbs: bool = False


class BackupExportResponse(BaseModel):
    """Response after starting a backup export job."""

    job_id: str
    message: str


class BackupStatusResponse(BaseModel):
    """Status of a backup export job."""

    job_id: str
    status: str
    progress_pct: float
    message: str
    error: str | None = None
    download_filename: str | None = None


# ── Helpers ──────────────────────────────────────────────────────────


def _get_file_size(path: Path) -> int:
    """Return file size in bytes, 0 if not found."""
    try:
        return path.stat().st_size if path.exists() else 0
    except OSError:
        return 0


def _collect_sample_files(data_dir: Path) -> list[Path]:
    """Collect all sample DB files from the samples directory."""
    samples_dir = data_dir / "samples"
    if not samples_dir.exists():
        return []
    return sorted(samples_dir.glob("sample_*.db"))


def _collect_reference_files(data_dir: Path) -> list[Path]:
    """Collect existing reference DB files."""
    files = []
    for name in REFERENCE_DB_FILES:
        p = data_dir / name
        if p.exists():
            files.append(p)
    return files


def _has_running_backup() -> bool:
    """Check if a backup export is already running."""
    import sqlalchemy as sa

    from backend.db.connection import get_registry
    from backend.db.tables import jobs

    registry = get_registry()
    with registry.reference_engine.connect() as conn:
        row = conn.execute(
            sa.select(jobs.c.job_id).where(
                jobs.c.job_type == "backup_export",
                jobs.c.status.in_(["pending", "running"]),
            )
        ).fetchone()
    return row is not None


# ── GET /api/backup/estimate ─────────────────────────────────────────


@router.get("/estimate", response_model=BackupEstimateResponse)
async def backup_estimate() -> BackupEstimateResponse:
    """Estimate backup archive size.

    Returns raw file sizes (actual .tar.gz will be smaller due to compression).
    """
    settings = get_settings()
    data_dir = settings.data_dir

    sample_files = _collect_sample_files(data_dir)
    sample_bytes = sum(_get_file_size(f) for f in sample_files)

    config_path = data_dir / "config.toml"
    disclaimer_path = data_dir / ".disclaimer_accepted"
    config_bytes = _get_file_size(config_path) + _get_file_size(disclaimer_path)

    ref_files = _collect_reference_files(data_dir)
    reference_bytes = sum(_get_file_size(f) for f in ref_files)

    total_without = sample_bytes + config_bytes
    total_with = total_without + reference_bytes

    return BackupEstimateResponse(
        sample_bytes=sample_bytes,
        config_bytes=config_bytes,
        reference_bytes=reference_bytes,
        total_without_ref_bytes=total_without,
        total_with_ref_bytes=total_with,
        total_without_ref_mb=round(total_without / (1024 * 1024), 1),
        total_with_ref_mb=round(total_with / (1024 * 1024), 1),
        sample_count=len(sample_files),
        reference_db_count=len(ref_files),
    )


# ── POST /api/backup/export ──────────────────────────────────────────


@router.post("/export", response_model=BackupExportResponse)
async def backup_export(
    body: BackupExportRequest,
) -> BackupExportResponse:
    """Start a backup export as a background Huey task.

    Creates a .tar.gz archive containing sample DBs, config.toml,
    and optionally reference databases.
    """
    if _has_running_backup():
        raise HTTPException(
            status_code=409,
            detail="A backup export is already in progress.",
        )

    from backend.tasks.huey_tasks import create_backup_job, run_backup_export_task

    job_id = create_backup_job()
    run_backup_export_task(job_id, body.include_reference_dbs)

    logger.info(
        "backup_export_started",
        job_id=job_id,
        include_reference_dbs=body.include_reference_dbs,
    )

    return BackupExportResponse(
        job_id=job_id,
        message="Backup export started.",
    )


# ── GET /api/backup/status/{job_id} ──────────────────────────────────


@router.get("/status/{job_id}", response_model=BackupStatusResponse)
async def backup_status(job_id: str) -> BackupStatusResponse:
    """Check the status of a backup export job."""
    import sqlalchemy as sa

    from backend.db.connection import get_registry
    from backend.db.tables import jobs

    registry = get_registry()
    with registry.reference_engine.connect() as conn:
        row = conn.execute(
            sa.select(
                jobs.c.status,
                jobs.c.progress_pct,
                jobs.c.message,
                jobs.c.error,
            ).where(
                jobs.c.job_id == job_id,
                jobs.c.job_type == "backup_export",
            )
        ).fetchone()

    if row is None:
        raise HTTPException(status_code=404, detail=f"Backup job {job_id} not found.")

    # Extract download filename from message if complete
    download_filename = None
    if row.status == "complete" and row.message:
        # Message format: "Backup complete: <filename>"
        prefix = "Backup complete: "
        if row.message.startswith(prefix):
            download_filename = row.message[len(prefix) :]

    return BackupStatusResponse(
        job_id=job_id,
        status=row.status,
        progress_pct=row.progress_pct,
        message=row.message,
        error=row.error,
        download_filename=download_filename,
    )


# ── GET /api/backup/download/{filename} ──────────────────────────────


@router.get("/download/{filename}")
async def backup_download(filename: str) -> FileResponse:
    """Download a completed backup archive.

    Only serves files from the downloads directory matching
    the genomeinsight_backup_*.tar.gz pattern.
    """
    # Validate filename to prevent path traversal
    if "/" in filename or "\\" in filename or ".." in filename:
        raise HTTPException(status_code=400, detail="Invalid filename.")

    if not filename.startswith("genomeinsight_backup_") or not filename.endswith(".tar.gz"):
        raise HTTPException(status_code=400, detail="Invalid backup filename.")

    settings = get_settings()
    archive_path = (settings.downloads_dir / filename).resolve()

    # Defense-in-depth: ensure resolved path stays within downloads_dir
    if not str(archive_path).startswith(str(settings.downloads_dir.resolve())):
        raise HTTPException(status_code=400, detail="Invalid filename.")

    if not archive_path.exists():
        raise HTTPException(status_code=404, detail="Backup file not found.")

    return FileResponse(
        path=str(archive_path),
        filename=filename,
        media_type="application/gzip",
    )
